hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/cancelairdrops/TokenCancelAirdropDecoder.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.cancelairdrops;
+
+import static com.hedera.hapi.node.base.ResponseCodeEnum.INVALID_TOKEN_ID;
+import static com.hedera.hapi.node.base.ResponseCodeEnum.PENDING_AIRDROP_ID_LIST_TOO_LONG;
+import static com.hedera.node.app.service.contract.impl.utils.ConversionUtils.asTokenId;
+import static com.hedera.node.app.spi.workflows.HandleException.validateFalse;
+import static com.hedera.node.app.spi.workflows.HandleException.validateTrue;
+
+import com.esaulpaugh.headlong.abi.Address;
+import com.esaulpaugh.headlong.abi.Tuple;
+import com.hedera.hapi.node.base.AccountID;
+import com.hedera.hapi.node.base.NftID;
+import com.hedera.hapi.node.base.PendingAirdropId;
+import com.hedera.hapi.node.base.TokenID;
+import com.hedera.hapi.node.base.TokenType;
+import com.hedera.hapi.node.token.TokenCancelAirdropTransactionBody;
+import com.hedera.hapi.node.transaction.TransactionBody;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.HtsCallAttempt;
+import com.hedera.node.config.data.TokensConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.Arrays;
+import javax.inject.Inject;
+
+public class TokenCancelAirdropDecoder {
+
+    // Tuple indexes
+    private static final int TRANSFER_LIST = 0;
+    private static final int SENDER = 0;
+    private static final int RECEIVER = 1;
+    private static final int TOKEN = 2;
+    private static final int SERIAL = 3;
+    private static final int HRC_RECEIVER = 0;
+    private static final int HRC_SERIAL = 1;
+
+    @Inject
+    public TokenCancelAirdropDecoder() {
+        // Dagger2
+    }
+
+    public TransactionBody decodeCancelAirdrop(@NonNull final HtsCallAttempt attempt) {
+        final var call = TokenCancelAirdropTranslator.CANCEL_AIRDROP.decodeCall(attempt.inputBytes());
+        final var maxPendingAirdropsToCancel =
+                attempt.configuration().getConfigData(TokensConfig.class).maxAllowedPendingAirdropsToCancel();
+        validateFalse(((Tuple[]) call.get(0)).length > maxPendingAirdropsToCancel, PENDING_AIRDROP_ID_LIST_TOO_LONG);
+
+        final var transferList = (Tuple[]) call.get(TRANSFER_LIST);
+        final var pendingAirdrops = new ArrayList<PendingAirdropId>();
+        Arrays.stream(transferList).forEach(transfer -> {
```

Comment:
I like the use of the stream at L65 below ... so ... you could _collect_ into a list instead of adding to it (via map + collect rather than foreach) ... it's a matter of taste how far into streams you want to go but actually there's no advantage to `Arrays.stream(...).forEach(...)` than the basic for loop if you're just adding to the result list ...

Reply:
Yes, thanks. Reworked to map and collect